### DISCUSSION TITLE
Exception.WithData method

### DISCUSSION THREAD

**Discussion by Khaos66** [dotnet/runtime#65555] [Ideas]
Exceptions all have a `Data` property, that no one uses. Thats sad. Because there could be additional information about the context of the exception, that doesn't fit in the `Message` property or that should be readable by code and not be parsed from the `Message`.

I think part of the problem is that `Data` is hard to set, when throwing an `Exception`.
So why not adding a `WithData(string key, object value)` method, to handle this issue like so?
```C#
throw new ArgumentException($"Invalid value for {nameof(myParam)}").WithData("value", myParam);
```

**Comment by Khaos66** [comment]
And while we're waiting... 
This can easily be done with an extension method like this
```C#
public static class ExceptionExtensions
{
    public static void WithData(this Exception e, string key, object value) => e.Data[key] = value;
}
```

**Comment by ufcpp** [comment]
```cs
throw new ArgumentException($"Invalid value for {nameof(myParam)}") { Data = { { "value", myParam } };
```

**Reply by Khaos66** [reply]
Yes. But that's what I refered to as '`Data` is hard to set' ;)

**Comment by svick** [comment]
You can already use code like the following:

```c#
throw new ArgumentException($"Invalid value for {nameof(myParam)}") { Data = { ["value"] = myParam } };
```

While this is more complicated than the suggested `WithData`, I think the difference is not that much and I'm not convinced it would lead to a significantly wider adoption of `Data`.

**Reply by Khaos66** [reply]
That might be true
